README.md
# csv-data-wrapper
A csv data wrapper class

## features

This is a python class to fiddle with csv files.

It was developed to be able to transform a spreadsheet into a importable csv file for PrestaShop.

When receiving a product spreadsheet from someone, it can contains errors:
- blanks where they should not be
- delimiter characters in some text field
- inconsistent capitalization
- ...

Also, when importing product data in PrestaShop, you will need specially crafted fields for:
- Categories
- Features

This is the process:
- save the spreadsheet as (input) csv
- run a script that uses this csv data wrapper (eg. test-data-wrapper.py)
  - to read the input csv file
  - process all you need to process
  - write the output csv file
- import the newly generated (output) csv into PrestaShop

It can:
- read a csv file
- show fields (like a very basic sql select statement)
- show an individual record
- add fixed fields
- add counter field
- add combined field
- copy a field
- replace in field
- map data to a new structure
- write a csv file


releasenotes.md
v0.1
First working version


data_table.py
## A data wrapper class around a data dictionary from a csv kind of datafile
## Version 0.1

import os
import csv
import logging

class DataTable:

    def __init__(self, displayfield='id'):
        self.displayfield = displayfield
        self.fields = []
        self.records = []

    def load_csv(self, filename, limit):
        logging.info(f"=== Loading CSV from [{filename}]")
        header = []
        csv_dict = {}
        try:
            rowcount = 1
            with open(filename, "r") as csvfile:
                csvreader = csv.DictReader(csvfile, delimiter='|', quotechar='"')
                self.fields = csvreader.fieldnames
                logging.info(f"HEADERLINE = {self.fields}")
                for row in csvreader:
                    self.records.append(row)
                    logging.debug(f" -> line [{csvreader.line_num}] -- [{row}]")
                    if limit > 0 and rowcount >= limit:
                        break
                    else:
                        rowcount += 1
        except FileNotFoundError:
            logging.error(f"ERROR - dude, file [{filename}] was not found")
        logging.info(f"RECORDS ===> [{rowcount}] data lines read from [{filename}]")
        return

    def show_fields(self, fieldnames=None):
        if fieldnames == None:
            fieldnames = self.fields
        logging.info(f"=== Showing fields {fieldnames}")
        logging.info(f" [HEADER]--[{']-['.join(fieldnames)}]")
        rowcount = 0
        for row in self.records:
            rowcount = rowcount + 1
            logging.debug(row)
            fieldvalues = []
            for fieldname in fieldnames:
                if fieldname in row:
                    fieldvalues.append(row[fieldname].strip())
                else:
                    fieldvalues.append('<' + fieldname.upper() + '>')
            logging.info(f" [#{str(rowcount)}]--[{']-['.join(fieldvalues)}]")
        return

    def show_record(self, id):
        logging.info(f"=== Showing record {id}")
        fieldnames = self.fields
        for record in self.records:
            if record[self.displayfield] == id:
                logging.debug(record)
                for fieldname in fieldnames:
                    logging.info(f" [{fieldname}]=[{record[fieldname]}]")
        return

    def remove_records(self, removedict):
        logging.info(f"=== Removing records where {removedict}")
        removefieldnames = removedict.keys()
        if len(removefieldnames) == 0:
            logging.warning(f" WARNING - no remove filter specified - not removing anything")
            return
        for removefieldname in removefieldnames:
            if removefieldname in self.fields:
                # logging.info(f" OK - remove field {removefieldname} exist")
                pass
            else:
                logging.error(f" ERROR - remove field {removefieldname} does not exist - not removing anything")
                return
        retainedrecords = []
        for row in self.records:
            logging.debug(row)
            # matchvalues = []
            matchcount = 0
            for fieldname in removefieldnames:
                # matchvalues.append(row[fieldname])
                if removedict[fieldname] == row[fieldname]:
                    matchcount += 1
            if matchcount == len(removefieldnames):
                logging.info(f" Exact match found for record [{row[self.displayfield]}] -- will be removed")
            else:
                # logging.info(f" Retaining [{row[self.displayfield]}]")
                retainedrecords.append(row)
            self.records = retainedrecords
        return

    def add_fixed_field(self, newfield, value):
        logging.info(f"=== Adding fixed field [{newfield}] - with value [{value}]")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FIELD!!")
        else:
            self.fields.append(newfield)
            for record in self.records:
                logging.debug(f"-- adding [{record[self.displayfield]}] - [{newfield}]=[{value}]")
                record[newfield] = str(value)
        return

    def add_counter_field(self, newfield, initialvalue):
        logging.info(f"=== Adding counter field [{newfield}] - starting at {initialvalue}")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FIELD!!")
        else:
            self.fields.append(newfield)
            countervalue = initialvalue
            for record in self.records:
                logging.debug(f"-- adding to [{record[self.displayfield]}] - [{newfield}]=[{countervalue}]")
                record[newfield] = str(countervalue)
                countervalue += 1
        return

    def add_combined_field(self, newfield, fieldstocombine, delimiter=''):
        logging.info(f"=== Adding combined field [{newfield}] - combining {fieldstocombine}")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FIELD!!")
        else:
            self.fields.append(newfield)
            for record in self.records:
                combinedfields = []
                for field in fieldstocombine:
                    if field in record.keys():
                        if record[field].strip() == '':
                            logging.warning(f"WARNING - WARNING - skipping empty field [{field}]")
                        else:
                            combinedfields.append(record[field].strip())
                    else:
                        logging.error(f"ERROR - ERROR - field [{field}] not found in this record - CANNOT COMBINE!!")
                record[newfield] = delimiter.join(combinedfields)
                logging.debug(f"-- adding to [{record[self.displayfield]}] - [{newfield}]=[{record[newfield]}]")
        return

    def add_combined_features_field(self, newfield, featurefields):
        logging.info(f"=== Adding combined features field [{newfield}] - combining {featurefields}")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FIELD!!")
        else:
            self.fields.append(newfield)
            for record in self.records:
                combinedfeatures = ''
                for featurefield in featurefields:
                    if featurefield in record.keys():
                        # Moeilijkheid:gemakkelijk:1:1
                        combinedfeatures += featurefield + ':' + record[featurefield].strip().capitalize().replace(",", ".") + ':1:0,'
                    else:
                        logging.error(f"ERROR - ERROR - feature field [{featurefield}] not found in this record - CANNOT COMBINE!!")
                record[newfield] = combinedfeatures.strip(',')
                logging.debug(f"-- adding to [{record[self.displayfield]}] - [{newfield}]=[{record[newfield]}]")
        return

    def add_combined_categories_field(self, newfield, categoryfields):
        logging.info(f"=== Adding combined categories field [{newfield}] - combining {categoryfields}")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FIELD!!")
        else:
            self.fields.append(newfield)
            for record in self.records:
                combinedcategories = ''
                for categoryfield in categoryfields:
                    if categoryfield in record.keys():
                        combinedcategories += ',' + record[categoryfield].strip().replace("/", "!")
                    else:
                        logging.error(f"ERROR - ERROR - category field [{categoryfield}] not found in this record - CANNOT COMBINE!!")
                record[newfield] = combinedcategories.strip(',')
                logging.debug(f"-- adding to [{record[self.displayfield]}] - [{newfield}]=[{record[newfield]}]")
        return

    def copy_field(self, sourcefield, newfield, prefix='', suffix=''):
        logging.info(f"=== Copying [{sourcefield}] into [{newfield}] with prefix=[{prefix}] and suffix=[{suffix}]")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FIELD!!")
        else:
            self.fields.append(newfield)
            for record in self.records:
                if sourcefield in record:
                    record[newfield] = prefix + record[sourcefield] + suffix
                    logging.debug(f"-- adding to [{record[self.displayfield]}] -> copied [{record[sourcefield]}] to [{record[newfield]}]")
                else:
                    record[newfield] = ""
                    logging.debug(f"-- WARNING - WARNING - No [{sourcefield}] found for [{record[self.displayfield]}] -- adding empty field")
        return

    def add_existing_image_field(self, basefield, newfield, checklocation='./', outputlocation='https://domain.org/images/', suffix='.jpg'):
        logging.info(f"=== Adding existing images based on [{basefield}] into [{newfield}] with location=[{checklocation}] and suffix=[{suffix}]")
        if newfield in self.fields:
            logging.error(f"ERROR - ERROR - field [{newfield}] already exists - NOT ADDING FILE-EXISTS FIELD!!")
            return
        if not basefield in self.fields:
            logging.error(f"ERROR - ERROR - base field [{basefield}] does not exists - NOT ADDING FILE-EXISTS FIELD!!")
            return
        self.fields.append(newfield)
        for record in self.records:
            imagetocheck = checklocation + record[basefield] + suffix
            imagetoadd = outputlocation + record[basefield] + suffix
            if os.path.isfile(imagetocheck):
                logging.info(f"-- Image [{imagetocheck}] found for [{record[self.displayfield]}] -> added [{imagetoadd}]")
                record[newfield] = imagetoadd
            else:
                logging.info(f"-- WARNING - WARNING - Image [{imagetocheck}] not found for [{record[self.displayfield]}] -- adding empty field")
                record[newfield] = ""
        return

    def replace_in_field(self, field, frompart, topart):
        logging.info(f"=== Replacing [{frompart}] into [{topart}] in field [{field}]")
        if field in self.fields:
            for record in self.records:
                record[field] = record[field].replace(frompart, topart)
                logging.debug(f"-- after replacement in [{record[self.displayfield]}]: [{field}]=[{record[field]}]")
        else:
            logging.error(f"ERROR - ERROR - field [{field}] does not exists - NOT REPLACING IN FIELD!!")
        return

    def re_map_table(self, fieldmap, displayfield='id'):
        logging.info(f"=== Re-mapping the records for Prestashop import")
        remapped_table = DataTable(displayfield=displayfield)
        remapped_fieldlist = []
        for field in fieldmap.values():
            remapped_fieldlist.append(field)
        remapped_table.fields = remapped_fieldlist
        for record in self.records:
            remapped_record = {}
            for key in fieldmap.keys():
                mapped_key = fieldmap[key]
                if key in record:
                    remapped_record[mapped_key] = record[key].strip()
                else:
                    remapped_record[mapped_key] = "[__" + key.strip().upper() + "__]"
            remapped_table.records.append(remapped_record)
        return(remapped_table)

    def generate_csv_array(self, fieldnames=None, delimiter=';'):
        logging.info(f"=== Generating csv array")
        if fieldnames == None:
            fieldnames = self.fields
        csvlist = []
        headerline = delimiter.join(fieldnames)
        csvlist.append(headerline)
        logging.debug(f"{headerline}")
        for record in self.records:
            values = []
            for fieldname in fieldnames:
                if fieldname in record.keys():
                    values.append(record[fieldname].strip())
                else:
                    values.append('')
            valuesline = delimiter.join(values)
            logging.debug(valuesline)
            csvlist.append(valuesline)
        return(csvlist)

    def write_csv(self, filename, delimiter=';'):
        logging.info(f"=== Writing csv file into [{filename}]")
        headerline = delimiter.join(self.fields)
        file = open(filename, "w")
        file.write("%s\n" % headerline)
        for record in self.records:
            valuesline = delimiter.join(record.values())
            file.write("%s\n" % valuesline)
        file.close()
        return


if __name__ == '__main__':
    print("This is a python object - do not run me")


dimarco-csv-oo-prepare-products.py
#!/usr/bin/python3 -tt

### FYI - example input file 'test-info/Website2021.csv'
### FYI - header line of this example file: "Code";"Type";"Wijn";"Land";"Regio";" Prijs/fles ";"Beschrijving";"Description";"Website";"Pakketactie"
##
## Run like so: python3 dimarco-csv-oo-prepare-products.py --inputfile data/input/Website2021.csv --limit 5

import csv
import time
import argparse
import logging
# import pprint
from data_table import DataTable

def chapter(title):
    number_of_title_dashes = (132 - len(title)) // 2
    title_dashes = "-" * number_of_title_dashes
    # logging.info("-" * 132)
    logging.info(title_dashes + title + title_dashes)
    # logging.info("-" * 132)
    return

def configure():
    parser = argparse.ArgumentParser(description='Morph Excel-csv into Prestashop-products-csv.')
    parser.add_argument('--inputfile', required=True, help='input file')
    parser.add_argument('--limit', required=True, help='number of records to process (0 processes them all)')
    args = parser.parse_args()
    configuration = {}
    configuration['limit']          = int(args.limit)
    configuration['inputfile']      = args.inputfile
    configuration['stamp']          = time.strftime("%Y%m%d-%H%M%S")
    configuration['outputfile']     = 'data/output/generated-ProductsForPrestaShop-' + configuration['stamp'] + '.csv'
    configuration['imglocation']    = 'https://www.dimarco.be/testshop/img/dimarco/'
    configuration['taxid']          = '666'
    configuration['fixed-active']   = '1'
    configuration['featurelist']    = ('Type', 'Land', 'Regio')
    configuration['categorieslist'] = ('cat1', 'cat2')
    configuration['fieldmap'] = {
            'generated-id': "PRODUCT-ID",
            'Code': "REFERENCE",
            'Wijn': "NAME",
            ' Prijs/fles ': "PRICE-INC",
            'Beschrijving': "LONG-DESC",
            'combined-features': "FEATURES",
            'combined-categories': "CATEGORIES",
            'fixed-aantal': "STOCK",
            'fixed-active': "ACTIVE",
            'fixed-taxid': "TAX-ID",
            'generated-img-url': "IMG-URL",
            'generated-beschrijving-kort': "SHORT-DESC"
        }
    return configuration

def main():

    logging.basicConfig(level=logging.INFO)
    logging.info('=== NORMAL START ===' + '=' * 111)

    chapter(f"Process arguments - configure")
    conf = configure()
    # pprint.pprint(conf, sort_dicts=False, width=132, indent=10, depth=1)

    # First - read the CSV files into a csv-table
    dimarco_data = DataTable(displayfield='Code')
    chapter(f"Load Di Marco csv [{conf['inputfile']}]")
    dimarco_data.load_csv(conf['inputfile'], conf['limit'])
    dimarco_data.show_fields(('Code', 'Wijn'))

    # # Then - we add some fields
    # chapter(f"Add some working fields")

    # # Then - we correct some errors in Griets data - I'd rather do it here than in the excel - because I sometimes get new ones
    # chapter(f"Correct Di Marco csv data")
    # dimarco_corrections.append(['Fotos', 'Borduurpakket 029', 'Omschrijving', 'tuinshort', 'tuinshort bloemen en gieters'])

    # # Then - we filter out records that we do not need
    # chapter(f"Remove unwanted records")
    # # mergedkeydict = remove_records_where(mergedkeydict, {'Categorie1': 'Verkocht', 'Categorie2': 'Verkocht'})

    # # Then - we add some extra fields
    chapter(f"Add some extra fields")
    dimarco_data.add_counter_field('generated-id', initialvalue=10001)
    dimarco_data.add_fixed_field('fixed-taxid', conf['taxid'])
    dimarco_data.add_fixed_field('fixed-active', '1')
    dimarco_data.add_fixed_field('fixed-aantal', '1000')
    dimarco_data.copy_field(sourcefield='Code', newfield='generated-img-url', prefix=conf['imglocation'], suffix='.jpg')
    dimarco_data.copy_field(sourcefield='Beschrijving', newfield='generated-beschrijving-kort')
    dimarco_data.show_fields(('Code', 'Wijn', 'fixed-taxid', 'fixed-active', 'generated-img-url'))

    # # Then - we correct some possible shizzle (extra spaces, title-ize product name)
    # chapter(f"Correct some possible shizzle (extra spaces, title-ize product name)")
    # mergedkeydict = correct_some_shizzle(mergedkeydict, ('Prijs', 'Merk', 'Omschrijving', 'Artikelnr', 'Aantal'))

    # # Then - we replace some shizzle - image url..
    # chapter(f"Replace some shizzle")
    # dimarco_data.replace_in_field('Imgurl', config['testproductimglocation'], config['realproductimglocation'])

    # # Then - we replace the empty value for the specified field
    # chapter(f"Replace empty values")
    # dimarco_data.fill_empty_field('Inhoud', 'DMC zesdraad')

    # # Then - we generate the short and long description
    # chapter(f"Generate the (short) description")
    # mergedkeydict = add_combined_description_field(mergedkeydict, 'AddedDescriptionShort', {'Omschrijving': 'Handwerkpakket', 'Ontwerper': 'Ontwerp', 'Leeftijd': 'Leeftijd', 'Onderwerp': 'Onderwerp', 'Inhoud': 'Inhoud', 'Artikelnr': 'Artikelnr'})

    # # Then - we process missing image url...
    # chapter(f"Process missing image url")
    # process_missing_images(mergedkeydict, config['altimagesdirectory'], config['realproductimglocation'], 'Imgurl')

    # # Then - we generate a complex combined feature field from the fields describing a feature
    chapter(f"Generate complex combined feature field from the feature fields")
    dimarco_data.add_combined_features_field('combined-features', conf['featurelist'])
    dimarco_data.show_fields(('Code', 'Wijn', 'combined-features'))

    # # Then - we combine the category fields into one category (list)field
    chapter(f"Combine the category fields into one category (list)field")
    # dimarco_data.add_fixed_field('fixed-cat1', 'category1')
    # dimarco_data.add_fixed_field('fixed-cat2', 'category2')
    dimarco_data.add_combined_categories_field('combined-categories', conf['categorieslist'])
    dimarco_data.show_fields(('Code', 'Wijn', 'combined-categories'))

    chapter(f"Re-map field for Prestashop")
    prestashop_table = dimarco_data.re_map_table(conf['fieldmap'], displayfield='REFERENCE')

    chapter(f"Some tests - output to screen")
    # dimarco_data.show_fields(('generated-id', 'Code', 'Wijn', 'Type', 'Land', 'Regio'))
    # prestashop_table.show_fields(('PRODUCT-ID', 'REFERENCE', 'NAME', 'FEATURES'))

    chapter(f"Generate resulting csv")
    # logging.info(f'====================================== BEGIN RESULTING CSV ======================================')
    # resultcsv = prestashop_table.generate_csv_array(delimiter='||')
    # logging.info(f'======================================= END RESULTING CSV =======================================')
    # print(resultcsv)

    # # Finally we write the resulting products csv file
    chapter(f"Writing resulting csv - [{conf['outputfile']}]")
    prestashop_table.write_csv(conf['outputfile'], delimiter=';')
    logging.info('=== NORMAL END ===' + '=' * 113)

    exit(0)


if __name__ == '__main__':
    main()


test-data-wrapper.py
#!/usr/bin/python3 -tt

## Run like so: python3 test-data-wrapper.py --inputfile data/input/test-date.csv --limit 5
## test-data header: "Active"|"Name"|"Category"|"Type"|"Country"|"Description"

import csv
import time
import argparse
import logging
from data_table import DataTable

def chapter(title, filler='-'):
    number_of_title_dashes = (132 - len(title)) // 2
    title_dashes = filler * number_of_title_dashes
    logging.info(title_dashes + title + title_dashes)
    return

def configure():
    parser = argparse.ArgumentParser(description='Morph Excel-csv into Prestashop-products-csv.')
    parser.add_argument('--inputfile', required=True, help='input file')
    parser.add_argument('--limit', required=True, help='number of records to process (0 processes them all)')
    args = parser.parse_args()
    configuration = {}
    configuration['limit']          = int(args.limit)
    configuration['inputfile']      = args.inputfile
    configuration['stamp']          = time.strftime("%Y%m%d-%H%M%S")
    configuration['outputfile']     = 'data/output/test-output-' + configuration['stamp'] + '.csv'
    # configuration['imglocation']    = 'https://www.dimarco.be/webshop/img/dimarcowines/'
    configuration['featurelist']    = ('Type', 'Country')
    configuration['categorieslist'] = ['fix-maincat', 'Category']
    configuration['finalfieldmap'] = {
            'gen-id': "UNIQUE-ID",
            'Active': "ACTIVE",
            'Name': "NAME",
            'Type': "TYPE",
            'Country': "COUNTRY",
            'Description': "DESCRIPTION",
            'fix-maincat': "CATEGORY",
            'Category': "SUBCATEGORY",
            'gen-short-description': "SHORT-DESCRIPTION",
            'gen-subtitle': "SUBTITLE",
            'combined-categories': "CATEGORIES",
            'combined-features': "FEATURES"
        }
    return configuration

def main():

    logging.basicConfig(level=logging.INFO)
    chapter(f" NORMAL START ", filler='=')

    chapter(f"Process arguments - configure")
    conf = configure()

    chapter(f"Load csv [{conf['inputfile']}]")
    input_data = DataTable(displayfield='Name')
    input_data.load_csv(conf['inputfile'], conf['limit'])
    # input_data.show_fields(('Active', 'Name', 'Description'))
    input_data.show_fields()

    chapter(f"Remove unwanted records")
    input_data.remove_records({'Active': 'N'})

    chapter(f"Replace some shizzle - newlines, semicolons, etc")
    # input_data.replace_in_field(field='Description', frompart='\n', topart='<br/>')
    input_data.replace_in_field(field='Description', frompart=';', topart='.')
    # input_data.show_fields(('Active', 'Name', 'Description'))

    chapter(f"Add some extra fields")
    input_data.add_counter_field('gen-id', initialvalue=10001)
    input_data.copy_field(sourcefield='Description', newfield='gen-short-description')
    input_data.add_fixed_field('fix-text-1', 'from')
    input_data.add_combined_field(newfield='gen-subtitle', fieldstocombine=['Type', 'fix-text-1', 'Country'], delimiter=' ')

    chapter(f"Generate complex combined feature field from the feature fields")
    input_data.add_combined_features_field('combined-features', conf['featurelist'])
    # input_data.show_fields(('Name', 'Product', 'combined-features'))

    chapter(f"Combine the category fields into one category (list)field")
    input_data.add_fixed_field('fix-maincat', 'Inhabitants')
    input_data.add_combined_categories_field('combined-categories', conf['categorieslist'])
    # input_data.show_fields(('Name', 'Product', 'combined-categories'))

    chapter(f"Re-map fields for Prestashop")
    output_data = input_data.re_map_table(conf['finalfieldmap'], displayfield='NAME')

    chapter(f"Some tests - output to screen")
    output_data.show_fields(('UNIQUE-ID', 'NAME', 'TYPE', 'CATEGORIES', 'FEATURES'))
    output_data.show_record('Bollie')

    # Finally we write the resulting products csv file
    chapter(f"Writing resulting csv - [{conf['outputfile']}]")
    output_data.write_csv(conf['outputfile'], delimiter=';')
    chapter(f" NORMAL END ", filler='=')

    exit(0)


if __name__ == '__main__':
    main()
